zson/lexer.go

```diff
@@ -0,0 +1,299 @@
+	cursor      []byte
+	primitive   *regexp.Regexp
+	indentation *regexp.Regexp
+}
+
+func NewLexer(r io.Reader) (*Lexer, error) {
+	// XXX Slurping in the entire file faciliated the implementation of
+	// the scanner logic here.  Now that the design is fleshed out, we need
+	// tweak things here to read from the input as a stream.  See issue #1802.
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	primitive := regexp.MustCompile(primitiveRE)
+	primitive.Longest()
+	indentation := regexp.MustCompile(indentationRE)
+	indentation.Longest()
+	return &Lexer{
+		buffer:      b,
+		cursor:      b,
+		primitive:   primitive,
+		indentation: indentation,
+	}, nil
+}
+
+func (l *Lexer) skip(n int) {
+	l.cursor = l.cursor[n:]
+}
+
+func (l *Lexer) peek() (byte, error) {
+	if len(l.cursor) > 1 {
+		return l.cursor[0], nil
+	}
+	return 0, io.EOF
+}
+
+func (l *Lexer) match(b byte) (bool, error) {
+	if err := l.skipSpace(); err != nil {
+		return false, err
+	}
+	return l.matchTight(b)
+}
+
+func (l *Lexer) matchTight(b byte) (bool, error) {
+	if len(l.cursor) > 1 {
+		ok := b == l.cursor[0]
+		if ok {
+			l.skip(1)
+		}
+		return ok, nil
+	}
+	return false, io.EOF
+}
+
+func (l *Lexer) matchBytes(b []byte) (bool, error) {
+	if err := l.skipSpace(); err != nil {
+		return false, err
+	}
+	return l.matchBytesTight(b)
+}
+
+func (l *Lexer) matchBytesTight(b []byte) (bool, error) {
+	if len(l.cursor) == 0 {
+		return false, io.EOF
+	}
+	n := len(b)
+	if n > len(l.cursor) {
+		return false, io.EOF
+	}
+	ok := bytes.Equal(b, l.cursor[:n])
+	if ok {
+		l.skip(n)
+	}
+	return ok, nil
+}
+
+func (l *Lexer) skipTo(b byte) ([]byte, error) {
+	for off := 0; off < len(l.cursor); off++ {
+		c := l.cursor[off]
+		if c == b {
+			return l.cursor[:off+1], nil
+		}
+	}
+	return nil, nil
+}
+
+func (l *Lexer) readByte() (byte, error) {
+	if len(l.cursor) == 0 {
+		return 0, io.EOF
+	}
+	b := l.cursor[0]
+	l.cursor = l.cursor[1:]
+	return b, nil
+}
+
+func (l *Lexer) peekRune() (rune, int, error) {
+	r, n := utf8.DecodeRune(l.cursor)
+	return r, n, nil
+}
+
+func (l *Lexer) peekRuneAt(n int) (rune, int, error) {
+	if len(l.cursor) < n {
+		return 0, 0, io.EOF
+	}
+	r, n := utf8.DecodeRune(l.cursor[n:])
+	return r, n, nil
+}
+
+var slashCommentStart = []byte("//")
+var starCommentStart = []byte("/*")
+
+func (l *Lexer) skipSpace() error {
+	for {
+		r, n, err := l.peekRune()
+		if err != nil {
+			return err
+		}
+		if unicode.IsSpace(r) {
+			l.skip(n)
+			continue
+		}
+		if r == '/' {
+			ok, err := l.matchBytesTight(slashCommentStart)
+			if err != nil {
+				return err
+			}
+			if ok {
+				if err := l.skipLine(); err != nil {
+					return err
+				}
+				continue
+			}
+			ok, err = l.matchBytesTight(starCommentStart)
+			if err != nil {
+				return err
+			}
+			if ok {
+				if err := l.skipMultiLine(); err != nil {
+					return err
+				}
+				continue
+			}
+		}
+		return nil
+	}
+}
+
+func isNewline(r rune) bool {
+	// from stackoverflow
```

Comment:
http://www.unicode.org/versions/Unicode13.0.0/ch05.pdf#G10213 is a better reference.